lib/Dialect/ESI/ESIPasses.cpp

```diff
@@ -84,13 +109,178 @@ void ESIToPhysicalPass::runOnOperation() {
     signalPassFailure();
 };
 
-} // end anonymous namespace.
+//===----------------------------------------------------------------------===//
+// Lower to RTL/SV conversions and pass.
+//===----------------------------------------------------------------------===//
+
+namespace {
+/// Assist the lowering steps for conversions which need to create auxiliary IR.
+class ESIRTLBuilder : public OpBuilder {
+public:
+  ESIRTLBuilder(Operation *top);
+
+  circt::rtl::RTLExternModuleOp declareStage();
+
+  // A bunch of constants for use in various places below.
+  const StringAttr a, a_valid, a_ready, x, x_valid, x_ready;
+  const StringAttr clk, rstn;
+  const Identifier parameters, WIDTH;
+
+private:
+  circt::rtl::RTLExternModuleOp declaredStage;
+  const Location loc;
+};
+} // anonymous namespace
+
+ESIRTLBuilder::ESIRTLBuilder(Operation *top)
+    : OpBuilder(top->getContext()), a(StringAttr::get("a", getContext())),
+      a_valid(StringAttr::get("a_valid", getContext())),
+      a_ready(StringAttr::get("a_ready", getContext())),
+      x(StringAttr::get("x", getContext())),
+      x_valid(StringAttr::get("x_valid", getContext())),
+      x_ready(StringAttr::get("x_ready", getContext())),
+      clk(StringAttr::get("clk", getContext())),
+      rstn(StringAttr::get("rstn", getContext())),
+      parameters(Identifier::get("parameters", getContext())),
+      WIDTH(Identifier::get("WIDTH", getContext())), declaredStage(nullptr),
+      loc(UnknownLoc::get(getContext())) {
+
+  auto regions = top->getRegions();
+  if (regions.size() == 0) {
+    top->emitError("ESI RTL Builder needs a region to insert RTL.");
+  }
+  auto &region = regions.front();
+  if (!region.empty())
+    setInsertionPoint(&region.front(), region.front().begin());
+}
+
+/// Write an 'ExternModuleOp' to use a hand-coded SystemVerilog module. Said
+/// module implements pipeline stage, adding 1 cycle latency. This particular
+/// implementation is double-buffered and fully pipelines the reverse-flow ready
+/// signal.
+circt::rtl::RTLExternModuleOp ESIRTLBuilder::declareStage() {
+  using namespace circt::rtl;
+
+  if (declaredStage)
+    return declaredStage;
+
+  auto name = StringAttr::get("ESI_PipelineStage", getContext());
+  // Since this module has parameterized widths on the a input and x output,
+  // give the extern declation a None type since nothing else makes sense.
+  // Will be refining this when we decide how to better handle parameterized
+  // types and ops.
+  SmallVector<ModulePortInfo, 8> ports = {
```

Comment:
For some reason, using `ArrayRef` here (and below) caused a segfault in the release build. Can someone enlighten me as to what's going on there?

Reply:
Array Ref is a pointer + length.  If you initialize it with a static array, the memory gets deallocated at the end of the statement.  I would use a C array, like:

`ModulePortInfo ports[] = { ... };`

C++ is a terrible terrible language btw ;-)

Reply:
If you don't name the variable, instead writing it like this:
```
declaredStage = create<RTLExternModuleOp>(loc, name, {{clk, PortDirection:INPUT, getI1Type(), 0}, 
  {...}
  ...});
return declaredStage;
```
then the storage should be valid as long as it's needed (presumably 'create' doesn't stash any pointers to this array in its return value - if it did, even the SmallVector solution wouldn't be sufficient (since 'declaredStage' outlives the SmallVector))

Reply:
Yep, that'll also work!

Reply:
Yep -- that worked. Thanks!